compiler/src/iree/compiler/InputConversion/TOSA/TosaToLinalgExt.cpp

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Convert TOSA operations to their LinalgExt equivalent.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtDialect.h"
+#include "iree-dialects/Dialect/LinalgExt/IR/LinalgExtOps.h"
+#include "iree/compiler/InputConversion/TOSA/PassDetail.h"
+#include "iree/compiler/InputConversion/TOSA/Passes.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/Dialect/Tosa/Transforms/Passes.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+using namespace mlir::tosa;
+
+namespace mlir {
+namespace iree_compiler {
+
+static SmallVector<utils::IteratorType> getNParallelLoopsAttrs(
+    unsigned nParallelLoops) {
+  return SmallVector<utils::IteratorType>(nParallelLoops,
+                                          utils::IteratorType::parallel);
+}
+
+// Converts tosa.scatter to the iree_linalg_ext.scatter operation. As the
+// LinalgExt version is not batched therefore we materialize the batch index
+// for each update.
+class ScatterConversion : public OpRewritePattern<tosa::ScatterOp> {
+ public:
+  using OpRewritePattern<tosa::ScatterOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tosa::ScatterOp op,
+                                PatternRewriter &rewriter) const final {
+    auto values = op.getValuesIn();
+    auto indices = op.getIndices();
+    auto updates = op.getInput();
+    auto valuesTy = values.getType().cast<ShapedType>();
+    auto indicesTy = indices.getType().cast<ShapedType>();
+    auto updatesTy = updates.getType().cast<ShapedType>();
+    ImplicitLocOpBuilder builder(op.getLoc(), rewriter);
+
+    if (!valuesTy.hasRank() || !indicesTy.getRank() || !updatesTy.getRank())
+      return rewriter.notifyMatchFailure(op,
+                                         "tosa.gather has unknown input rank");
+
+    // Materialize the unary index dimension.
+    llvm::SmallVector<int64_t> expandIndShape{indicesTy.getDimSize(0),
+                                              indicesTy.getDimSize(1), 1};
+    SmallVector<ReassociationExprs, 4> expandIndMap;
+    expandIndMap.push_back({
+        builder.getAffineDimExpr(0),
+    });
+    expandIndMap.push_back({
+        builder.getAffineDimExpr(1),
+        builder.getAffineDimExpr(2),
+    });
+
+    indices = builder.create<tensor::ExpandShapeOp>(
+        indicesTy.clone(expandIndShape), indices, expandIndMap);
+    indicesTy = indices.getType().cast<ShapedType>();
+
+    // Materialize the batch indice as LinalgExt scatter is not batched.
+    {
+      llvm::SmallVector<Value> dynDims;
+      for (int i = 0; i < indicesTy.getRank(); i++)
```

Comment:
Ah ha, I did misread in the LLVM review, but do see the check inside conditional here :)

Reply:
Fixed the iterator hear and below.